### ISSUE TITLE
Failed to find patch to apply to Voicemeeter

### ISSUE THREAD

**Issue by kazuviking2** [iagocq/voicemeeter-activator#4]
3.1.1.1 version. Doesnt matter what i do it fails to patch the exe.

**Comment by iagocq** [comment]
what error message shows up?

**Comment by iagocq** [comment]
Also, please give information about what you used to build the program, like MinGW gcc version, how you installed it (through MSYS2 or another installer)

**Comment by kazuviking2** [comment]
TLDR only w64devkit worked on my pc.

The things i tried so far, MSYS2 MinGW64, w64devkit and sourceforge MinGW64 all of the latest version. The sourceforge version had issue with the GetNativeSytemInfo. MSYS2 UCRT64 and MinGW64 patch issue. After more trial and error only the [w64devkit](https://github.com/niXman/mingw-builds-binaries/releases) worked but that only if i compiled it in the VB folder for some reason.
As a sanity check i tried compiling on my laptop lo and behold it just worked with MSYS2. 
A last question if you don't mind. Can i buy the vaio extention without running into issues with the activator? Thx for your work.

**Comment by iagocq** [comment]
You're saying that it only works on your PC if you build the executable inside the VB folder with w64devkit? Even if you build it elsewhere and move the executable it doesn't work?

On your laptop, the other compiler (w64devkit gcc) fail just like other compilers failed on your PC? I'm assuming you compiled it with MSYS2 UCRT64.

Does the exe that you built on your laptop work on your PC?

Sorry for the load of questions! I'm trying to narrow the issue down.

I'm also currently trying to replicate the issue with different compilers in Windows Sandbox.

**Comment by kazuviking2** [comment]
So did more testing. On that laptop(laptop1) both compilers(w64devkit and UCRT64) worked first try(compiled in VB folder and in random folder). Its activator worked on the PC and on laptop2 without issues. Put a new drive and fresh win10 install in laptop2 resulted in the same patch issue i had with my pc. The sourceforge version couldn't comply on all 3, same getnativesystem info error. w64devkit worked on laptop2 first try(compiled outside VB), UCRT64 failed to patch but fixed itself after a few restarts(still don't know how or why). Now the confusing part, the activator compiled on the PC outside the VB folder works but only with w64devkit, UCRT64 still gives patch error(this is where i gave up).
I gave up testing more, w64devkit worked and it was the easiest to set up.
Some random info: PC and laptop2 is relatively fresh win10 install with only the necessary drivers. System was restarted after every environmental change(its not needed but hey). Compiled on L1 works in PC and L2, Compiled on L2 works in PC and L1. Compiled on PC only works in PC, patch error otherwise.

**Comment by iagocq** [comment]
Here are the results of my tests.

## Compilation

I used the Win32 threads versions of each following distribution.

What compiled without issues for me:
- MSYS2 UCRT64 (compiled to activator.ucrt64.exe)
- w64devkit x86_64 (from https://github.com/skeeto/w64devkit) (compiled to activator.w64devkit.64.exe)
- niXman's x86_64 MinGW (MSVCRT) (from https://github.com/niXman/mingw-builds-binaries) (compiled to activator.nixman.msvcrt.64.exe)
- niXman's x86_64 MinGW (UCRT) (compiled to activator.nixman.ucrt.64.exe)

What compiled without critical issues for me:
- Sourceforge MinGW-W64 x86_64 (compiled to activator.sf.64.exe)
- Sourceforge MinGW-W64 i686 (compiled to activator.sf.32.exe)
- w64devkit i686 (compiled to activator.w64devkit.32.exe)
- niXman's i686 MinGW (MSVCRT) (compiled to activator.nixman.msvcrt.32.exe)
- niXman's i686 MinGW (UCRT) (compiled to activator.nixman.ucrt.32.exe)

For the x86_64 Sourceforge versions, I got a warning about an implicit declaration of Wow64GetThreadContext, but this is not really a problem because the activator was built successfully.

For the i686 versions, I got a warning about an integer to pointer cast of different size (expected because I'm casting an uint64 to a 32 bit pointer, I should be using size_t instead)

## Activation

All x86_64 executables built in the Windows Sandbox worked without issues against voicemeeter8x64.exe versions 3.1.1.1 and 3.0.2.8.

All executables built in the Windows Sandbox worked without issues against voicemeeter8.exe version 3.0.2.8.

**Comment by iagocq** [comment]
In conclusion, I couldn't replicate the GetNativeSystemInfo. Could you give the link to the exact MinGW build you used that had that issue? It's not high priority, because the outer function is not used at all, so you can remove it without affecting the patching behaviour.

I think the patch issue might be because I'm not properly calculating the offset of the target function to be patched in memory. When first writing this activator, by experimenting I found an offset (`0xc00`) that got me from the offset where the function is in the executable file to the memory address where the function was loaded. It might be that this value is not correct for all cases. I'll try to find a correct way to calculate the offset in a way that doesn't depend on this honestly hacky magic value.

**Comment by kazuviking2** [comment]
No wonder it failed to compile, the sourceforge version that i downloaded(very first result in Duck Duck Go when i searched for MinGW) is from 2013. This was my fault for not checking but i only learned the existance of MinGW yesterday.
![image](https://github.com/iagocq/voicemeeter-activator/assets/79218956/8f935eef-58a4-4376-a466-4af6baae40ee)

**Comment by iagocq** [comment]
I removed the magic calculation that gave me the address of the function to be patched. I'm still not sure it will fix your issue (the magic number should not change from machine to machine), but no harm in trying.

@kazuviking2 could you try the activator again?

**Comment by iagocq** [comment]
also, what version of Windows 10 you installed on Laptop 2 that caused the activator to not work? When I get the chance, I'll try to run the same version in a fresh VM and see if I can replicate your issue.

**Comment by kazuviking2** [comment]
The new one works with UCRT64 on the PC, patches witout issue. Too late to boot up laptop 2 so that will have to wait.  Windows version is 19045.3803, outdated version but this is what i had on hand. 
I doubt anyone will try to use a fresh win10 install but you never know.

**Comment by iagocq** [comment]
I was unable to replicate the issue on a fresh Windows 10 19045.3803 install on a VM.

I installed MSYS2 UCRT64 MinGW, compiled the activator from commit 8c9663c091f52f805ead43cb6753a21dbd7c80a6, installed Voicemeeter 3.1.1.1, then ran the activator with no issue.

I don't think I'll be able to replicate it on my machine, so I won't be able to know for sure if the issue is fixed. Let me know if the activator is working correctly on all your machines.
